substrate/frame/revive/src/precompiles/custom/xcm.rs

```diff
@@ -0,0 +1,113 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::Origin;
+use crate::{
+	precompiles::{AddressMatcher, Error, Ext, ExtWithInfo, Precompile},
+	Config,
+};
+use alloc::vec::Vec;
+use alloy_core::sol;
+use codec::{DecodeAll, Encode};
+use core::{marker::PhantomData, num::NonZero};
+use xcm_builder::{ExecuteController, ExecuteControllerWeightInfo, SendController};
+pub use IXcm::IXcmCalls;
+
+pub struct XcmPrecompile<T>(PhantomData<T>);
+
+sol! {
+	/// @title Defines all functions that can be used to interact with XCM
+	/// @author Tiago Bandeira
+	/// @dev Parameters MUST use SCALE codec serialisation
+	interface IXcm {
+		/// @notice Execute an XCM message locally with the caller's origin
+		/// @param message The XCM message to send
+		function xcmExecute(bytes calldata message) external;
+
+		/// @notice Send an XCM message to a destination chain
+		/// @param destination The destination location, encoded according to the XCM format
+		/// @param message The XCM message to send
+		function xcmSend(bytes calldata destination, bytes calldata message) external;
+	}
+}
+
+impl<T: Config> Precompile for XcmPrecompile<T> {
+	type T = T;
+	const MATCHER: AddressMatcher = AddressMatcher::Fixed(NonZero::new(10).unwrap());
+	const HAS_CONTRACT_INFO: bool = false;
+	type Interface = IXcm::IXcmCalls;
+
+	fn call(
+		_address: &[u8; 20],
+		input: &Self::Interface,
+		env: &mut impl Ext<T = Self::T>,
+	) -> Result<Vec<u8>, Error> {
+		let origin = env.origin();
```

Comment:
What is in `env.origin`? The address of the contract or of the user calling the contract?

Reply:
So `env.caller()` will be whoever called **this** particular contract?

Reply:
Exactly.